src/object.d

```diff
@@ -3241,6 +3241,269 @@ template RTInfo(T)
+{
+    import core.internal.traits : Unqual;
+    alias U1 = Unqual!T1;
+    alias U2 = Unqual!T2;
+
+    // objects
+    // Old code went to object.d/TypeInfo_array.compare and then to
+    // object.d/TypeInfo_Class.compare
+    // which combined result in the same code as below.
+    static if (is(U1 : Object) && is(U2 : Object))
+    {
+        auto c  = cast(sizediff_t)(s1.length - s2.length);
+        if (c == 0)
+        {
+            for (size_t u = 0; u < s1.length; u++)
+            {
+                auto o1 = () @trusted { return s1.ptr[u]; }();
+                auto o2 = () @trusted { return s2.ptr[u]; }();
+
+                if (o1 is o2)
+                    continue;
+
+                // Regard null references as always being "less than"
+                if (o1)
+                {
+                    if (!o2)
+                        return 1;
+                    c = o1.opCmp(o2);
+                    if (c != 0)
+                        break;
+                }
+                else
+                {
+                    return -1;
+                }
+            }
+        }
+        return c < 0 ? -1 : (c > 0);
+    }
+
+    // floating point types
+    else static if (__traits(isFloating, U1))
+    {
+        static if (is(U1 == ifloat)) alias F = float;
+        else static if (is(U1 == idouble)) alias F = double;
+        else static if (is(U1 == ireal)) alias F = real;
+        else alias F = U1;
+
+        return () @trusted { return Array1!F.compare(cast(F[])s1, cast(F[])s2); }();
+    }
+
+    // char types = > dstrcmp
+    else static if ((is(U1 == ubyte) && is(U2 == ubyte))
+        || (is(U1 == void) && is(U2 == void))
+        || (is(U1 == bool) && is(U2 == bool))
+        || (is(U1 == char) && is(U2 == char)))
+    {
+        if (!__ctfe)
+        {
+            import core.internal.string : dstrcmp;
+            return () @trusted { return dstrcmp(cast(char[])s1, cast(char[])s2); }();
+        }
+        else
+        {
+            // pretty ugly...
+            auto len = s1.length;
+
+            if (s2.length < len)
+                len = s2.length;
+
+            foreach (const u; 0 .. len)
+            {
+                auto e1 = () @trusted { return s1.ptr[u]; }();
+                auto e2 = () @trusted { return s2.ptr[u]; }();
+
+                if (e1 < e2)
+                    return -1;
+                else if (e1 > e2)
+                    return 1;
+            }
+
+            return s1.length < s2.length ? -1 : (s1.length > s2.length);
+        }
+    }
+
+    // integral
+    else
+    {
+        auto len = s1.length;
+
+        if (s2.length < len)
+            len = s2.length;
+
+        foreach (const u; 0 .. len)
+        {
+            auto e1 = () @trusted { return s1.ptr[u]; }();
+            auto e2 = () @trusted { return s2.ptr[u]; }();
+
+            if (e1 < e2)
+                return -1;
+            else if (e1 > e2)
+                return 1;
```

Comment:
Well, do what the old code did. Which I would guess would be: essentially for all types for which we can get an integer comparison result, we should do that - call `opCmp` for structs/classes and / `__cmp` (i.e. the template introduced here) for arrays.